File: ticTacToe.py
import random


board=["-", "-", "-",
       "-", "-", "-",
       "-", "-", "-"]
currentPlayer ="X"
winner =None
gameRunning =True

def printBoard(board):
    print(board[0]+" | "+board[1]+" | "+board[2])
    print("---------")
    print(board[3]+" | "+board[4]+" | "+board[5])
    print("---------")
    print(board[6]+" | "+board[7]+" | "+board[8])


def playerInput(board):
    inp = int(input("Select a spot 1-9: "))
    if inp>=1 and inp<=9 and board[inp-1]=="-":
        board[inp-1]= currentPlayer
    else:
        print("Oops invalid input")

def checkHorizontal(board):
    global winner
    if board[0]==board[1]==board[2] and board[0]!="-":
        winner=board[0]
        return True
    elif board[3]==board[4]==board[5] and board[3]!="-":
        winner=board[3]
        return True
    elif board[6]==board[7]==board[8] and board[6]!="-":
        winner=board[6]
        return True
    
def checkVertical(board):
    global winner
    if board[0]==board[3]==board[6] and board[0]!="-":
        winner=board[0]
        return True
    elif board[1]==board[4]==board[7] and board[1]!="-":
        winner=board[1]
        return True
    elif board[2]==board[5]==board[8] and board[2]!="-":
        winner=board[2]
        return True
    
def checkDiagonal(board):
    global winner
    if board[0]==board[4]==board[8] and board[0]!="-":
        winner=board[0]
        return True
    elif board[2]==board[4]==board[6] and board[2]!="-":
        winner=board[1]
        return True
    
def checkTie(board):
    global gameRunning
    if "-" not in board:
        printBoard(board)
        print("It's a tie boi")
        gameRunning = False

def switchPlayer():
    global currentPlayer
    if currentPlayer == "X":
        currentPlayer = "O"
    else:
        currentPlayer = "X"

def checkWin(board):
    global gameRunning
    if checkDiagonal(board) or checkHorizontal(board) or checkVertical(board):
        print(f"The Winner is {winner}!")
        gameRunning=False

def computer(board):
    while currentPlayer == "O":
        position = random.randint(0, 8)
        if board[position] == "-":
            board[position] = "O"
            switchPlayer()

while gameRunning:
    printBoard(board)
    playerInput(board)
    checkWin(board)
    checkTie(board)
    switchPlayer()
    computer(board)
    checkWin(board)
    checkTie(board)

File: ticTacToe.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>

using namespace std;
vector<char> board = { '-', '-', '-',
                            '-', '-', '-',
                            '-', '-', '-' };
char currentPlayer = 'X';
char winner = '\0';
bool gameRunning = true;

void printBoard(const vector<char>& board) {
    cout << board[0] << " | " << board[1] << " | " << board[2] << endl;
    cout << "---------" << endl;
    cout << board[3] << " | " << board[4] << " | " << board[5] << endl;
    cout << "---------" << endl;
    cout << board[6] << " | " << board[7] << " | " << board[8] << endl;
}

void playerInput(vector<char>& board) {
    int inp;
    cout << "Select a spot 1-9: ";
    cin >> inp;
    if (inp >= 1 && inp <= 9 && board[inp - 1] == '-') {
        board[inp - 1] = currentPlayer;
    }
    else {
        cout << "Oops invalid input" << endl;
    }
}

bool checkHorizontal(const vector<char>& board) {
    if (board[0] == board[1] && board[1] == board[2] && board[0] != '-') {
        winner = board[0];
        return true;
    }
    else if (board[3] == board[4] && board[4] == board[5] && board[3] != '-') {
        winner = board[3];
        return true;
    }
    else if (board[6] == board[7] && board[7] == board[8] && board[6] != '-') {
        winner = board[6];
        return true;
    }
    return false;
}

bool checkVertical(const vector<char>& board) {
    if (board[0] == board[3] && board[3] == board[6] && board[0] != '-') {
        winner = board[0];
        return true;
    }
    else if (board[1] == board[4] && board[4] == board[7] && board[1] != '-') {
        winner = board[1];
        return true;
    }
    else if (board[2] == board[5] && board[5] == board[8] && board[2] != '-') {
        winner = board[2];
        return true;
    }
    return false;
}

bool checkDiagonal(const vector<char>& board) {
    if (board[0] == board[4] && board[4] == board[8] && board[0] != '-') {
        winner = board[0];
        return true;
    }
    else if (board[2] == board[4] && board[4] == board[6] && board[2] != '-') {
        winner = board[2];
        return true;
    }
    return false;
}

bool checkTie(const vector<char>& board) {
    for (char spot : board) {
        if (spot == '-') {
            return false;
        }
    }
    return true;
}

void switchPlayer() {
    if (currentPlayer == 'X') {
        currentPlayer = 'O';
    }
    else {
        currentPlayer = 'X';
    }
}

void checkWin(const vector<char>& board) {
    if (checkDiagonal(board) || checkHorizontal(board) || checkVertical(board)) {
        cout << "The Winner is " << winner << "!" << endl;
        gameRunning = false;
    }
}

void computer(vector<char>& board) {
    while (currentPlayer == 'O') {
        int position = rand() % 9;
        if (board[position] == '-') {
            board[position] = 'O';
            switchPlayer();
        }
    }
}

int main() {
    srand(time(nullptr));
    
    while (gameRunning) {
        printBoard(board);
        playerInput(board);
        checkWin(board);
        if (checkTie(board)) {
            printBoard(board);
            cout << "It's a tie boi" << endl;
            gameRunning = false;
            break;
        }
        switchPlayer();
        computer(board);
        checkWin(board);
        if (checkTie(board)) {
            printBoard(board);
            cout << "It's a tie boi" << endl;
            gameRunning = false;
        }
    }

    return 0;
}
